src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java

```diff
@@ -412,15 +433,15 @@ private UnfilteredRowIterator applyIndexFilter(PrimaryKey key, UnfilteredRowIter
             {
                 queryContext.rowsFiltered++;
 
-                if (tree.isSatisfiedBy(key.partitionKey(), staticRow, staticRow))
+                if (tree.isSatisfiedBy(partition.partitionKey(), staticRow, staticRow))
                     hasMatch = true;
             }
 
             if (!hasMatch)
             {
                 // shadowed by expired TTL or row tombstone or range tombstone
                 if (topK)
-                    queryContext.vectorContext().recordShadowedPrimaryKey(key);
+                    queryContext.vectorContext().recordShadowedPrimaryKey(keys);
```

Comment:
I don't seem have broken any tests w/ this change, and I can only suspect that's because all the existing testing for Vector Search use single-row partitions?

Reply:
```
        private UnfilteredRowIterator applyIndexFilter(List<PrimaryKey> keys, UnfilteredRowIterator partition, FilterTree tree)
        {
            Row staticRow = partition.staticRow();
            List<Unfiltered> matchingRows = new ArrayList<>();
            boolean hasMatch = false;
            Set<PrimaryKey> keysToShadow = new HashSet<>(keys);

            // We need to filter the partition rows before filtering on the static row. If this is done in the other
            // order then we get incorrect results if we are filtering on a partition key index on a table with a
            // composite partition key.
            while (partition.hasNext())
            {
                Unfiltered unfiltered = partition.next();

                if (unfiltered.isRow())
                {
                    queryContext.rowsFiltered++;

                    if (tree.isSatisfiedBy(partition.partitionKey(), (Row) unfiltered, staticRow))
                    {
                        matchingRows.add(unfiltered);
                        hasMatch = true;

                        PrimaryKey key = keyFactory.create(partition.partitionKey(), ((Row) unfiltered).clustering());
                        keysToShadow.remove(key);
                    }
                }
            }

            if (!hasMatch)
            {
                queryContext.rowsFiltered++;

                if (tree.isSatisfiedBy(partition.partitionKey(), staticRow, staticRow))
                {
                    hasMatch = true;
                    keysToShadow.clear();
                }
            }

            // shadowed by expired TTL or row tombstone or range tombstone
            if (topK)
                queryContext.vectorContext().recordShadowedPrimaryKey(keysToShadow);

            if (!hasMatch)
            {
                // If there are no matches, return an empty partition. If reconciliation is required at the
                // coordinator, replica filtering protection may make a second round trip to complete its view
                // of the partition.
                return null;
            }

            // Return all matches found, along with the static row... 
            return new PartitionIterator(partition, staticRow, matchingRows.iterator());
        }
```
this might work.. but we need some tests...

Reply:
Yeah, that's more or less what I had imagined, modulo fixing it up to avoid allocations for the topK case, etc.

It's unfortunate that there are no tests that seem to cover this case, and I don't quite know where to start. Would `VectorUpdateDeleteTest` be a place to look?

Reply:
I'm actually not even sure if we need the explicit static filtering block after the `while` loop. That block was originally added before CASSANDRA-19034 as an optimization to short-circuit before the `while` loop. Removing it would make some things simpler. I'm going to try it out and push something...